src/utils.jl

```diff
@@ -77,6 +77,56 @@ function sideof(p::Point{2,T}, c::Chain{2,T}) where {T}
   ifelse(isapprox(w, zero(T), atol=atol(T)), :OUTSIDE, :INSIDE)
 end
 
+"""
+    sideof(point, mesh)
+
+Determines whether a `point` is inside, outside or on the surface of a `mesh`.
+Possible results are `:INSIDE`, `:OUTSIDE`, or `:ON`.
+
+### Notes
+
+Uses a ray-casting algorithm.
+"""
+function sideof(point::Point{3,T}, mesh::Mesh{3,T}) where {T}
+  if !(eltype(mesh) <: Triangle)
+    throw(ArgumentError(
+          "This function only works for surface meshes with triangles as elements."))
+  end
+
+  z = last.(coordinates.(extrema(mesh)))
+  r = Ray(point, Vec(zero(T), zero(T), 2 * (z[2] - z[1])))
+  v = vertices(mesh)
+
+  intersects = false
+  edgecrosses = 0
+  cornerscrossing = Point{3,T}[]
+  for e in mesh
+    result = intersection(r, e) do I
+      if type(I) == CrossingRayTriangle
+        intersects = !intersects
+      elseif type(I) ∈ (EdgeTouchingRayTriangle,
+                        CornerTouchingRayTriangle,
+                        TouchingRayTriangle)
+        return :ON
+      elseif type(I) == EdgeCrossingRayTriangle
+        edgecrosses += 1
+      elseif type(I) == CornerCrossingRayTriangle
+        id = findfirst(p -> p ≈ get(I), v)
+        if v[id] ∉ cornerscrossing
+            push!(cornerscrossing, v[id])
+            intersects = !intersects
+        end
+      end
+      return nothing
+    end
+    !isnothing(result) && return :ON
+  end
+
+  # check how many edges we crossed
+  isodd(round(Int, edgecrosses / 2)) && (intersects = !intersects)
+  intersects ? (return :INSIDE) : (return :OUTSIDE)
```

Comment:
can you write the English-equivalent of this algorithm? I feel that it could be simplified, it is currently not straightforward to digest. The variable names mixing single-letter variables with long-words variables also makes it difficult to follow. Maybe cornerscrossing should be a Set? You are constantly checking if the point belongs to the list and then adding it. Maybe the lines with `intersects = !intersects` should be more explicit with `true` and `false` instead?

Reply:
Sure, probably should have explained in English in the description of this PR anyways. The algorithm really is easy in the basic idea, it is the 'tricky' cases that make it confusing/difficult to handle. We just need to check how many intersections there are of a ray with the surface - if odd, the point is inside, if even, outside. This is why we can't set intersects to true or false because we don't keep track of how many intersections and are just flipping back and forth like even and odd. In our case, we also want to check if the point is on the surface, but that part is easy to see (lines 107-110).

The tricky parts are when the ray hits an edge or a corner of a triangle of the surface mesh.
1. If it hits an edge - we have to consider that an edge is shared by 2 triangles so when we reach the other triangle in the loop over all elements that intersection will get counted again. To handle this, I am just counting the total number of edge intersects (lines 111-112) and then divide by 2 to get the actual amount of intersections of edges (line 126). Now if this number is even, it "cancels" out but if odd, we count it as another intersection in the global tracking and "flip" the switch of `intersects` as if the ray simply intersected the middle of a triangle (line 126).
2. if the ray hit a corner of a triangle - this is the part where there is probably a better way to do this. Since the number of triangles that share a corner is not always the same like an edge, you can't just divide by that number at the end. There are 2 main ways I thought of to go about this:

- what I did here: when I do come across a ray hitting a corner, we check if that point has already been "hit" (if we are currently considering an element that shares the corner we already came across) (line 115) and if not, then we add it to the list of corners we have intersected (line 116) and flip the global tracker of intersects (line 117).
- each time we hit a corner, we check to see how many elements it shares the corner with via coboundary (lets call that N) and we can add N-1 to a sum (called A), and also add 1 to a sum counting how many corners (including duplicates from shared elemnts) we have hit (call it C) so that we can find how many unique corners we hit via C-A. add again if that number is odd then we flip the global `intersects`.

Number of shared elements for each corner:
c1 = 3
c2 = 4
c3 = 6

Example cases
hit all: 13-10=3 is odd -> flip global intersects switch
hit c1and c2: 7-5=2 is even -> do nothing (it "cancels")
hit c1 only: 3-2=1 is odd -> flip global intersects switch

I don't know the complexity of finding the number of shared elements via coboundary, so it could be better I don;t know. But my latest commit gets rid of the useless `findfirst` and instead compares approximately to the point the ray-triangle returns which should be the same as the vertex in the mesh up to numerical precision for that type. We can benchmark and go through all that, but I think it is pointless, because this algorithm is meant to be **robust**, not **performant**. if you really care about performance, there are much faster algorithms, but they are not as robust (they are approximate). For what its worth, I have implemented this approximate type of approach for my personal use case and I find it works well enough for me (but I am using medium-ish size surface meshes ~30k elements) and this algorithms accuracy scales with number of elements (not good for a box or something like that).

Reply:
See Fast winding numbers for soups and clouds, Gavin Barill, 2018. I implemented a naive (not as performant or accurate) approach of this myself. I am looking to implement this, but it is not completely foolproof like the algorithm we have here. ideally we have both and the user can choose what is more important to them.